### DISCUSSION TITLE
Alternative Number Support

### DISCUSSION THREAD

**Discussion by Taar** [convert-units/convert-units#333] [Ideas]
This discussion is a continuation of the conversation on issue #320 .

The [alternative-number-support](https://github.com/convert-units/convert-units/tree/alternative-number-support) branch contains the proposed feature implementation.

The goal of this feature are list below:
- Provide a way for the use of another number object like Decimal.js instead of a regular `number` (which would solve various floating point issues)
- Keeping the library dependency free
- Preserving backwards compatibility for methods and bundled definitions

Please let me know what your opinion of the proposed changes :smile:

**Comment by Taar** [comment]
I'm going to address these comments from the #320 issue

> @Taar, I want to highlight that the Decimal.js solution I suggested is fully backwards compatible, as the return type of any exported function remains unchanged; only internal changes were made. However, I prefer your solution because it avoids adding any dependencies.

Sorry, I should provided more details on the PR (#328). I really didn't like the fact that Decimal.js was required within the definitions to solve the floating point issue. I never liked that fact that there are expressions as part of the definitions to begin with! I wanted to find a way to remove these expressions altogether while also eliminating the definitions as a potential source of floating point errors. Even in the proposed changes, if a custom measure is defined that contains an expression, there is the possibility that the expression could be the source of a floating point error. The best way, imo, to combat this is to update the documentation with a floating point error section that highlights the dangers of using expressions in the definitions and provides ways to combat the issue while it being opt-in. While at the same time, not allowing (and discouraging) expressions within the definitions. With all that said, I thought that the changes to the definitions was a breaking change.

The other headache that comes from that is that if a user defines their own measure or extends a measure, they are likely going to be using a different version of Decimal.js which could cause a lot of hard to track down issues later down the road. There is one way to sort of solve this issue, which is to define Decimal.js as a peer dependency but I'm still not convinced it was the best approach.

> Regarding the Wrapper, I believe it may unnecessarily increase code complexity. I see no compelling reason to use anything but a number, though I'm open to counterarguments.

I think I understand what you're trying to say here but let me know if my assumption is incorrect. I would expect that someone that wants to use Decimal.js (for example) with this library is already using it within other parts of their application. Personally, I would find it annoying to have to convert a `Decimal` to a `number` or `string` in order to use it with this library. I also would like the methods to return a `Decimal` since converting it to a `number` or `string` and returning that value could cause a loss in precision. Though, I will admit that the current proposed changes do not allow a user to pass a `Decimal` object but that is something I'm going to add in once I have a free moment.

I also want the user to have the ability to use whatever they want! Maybe they have a special custom decimal object or maybe they prefer a different library. ECMAScript even has a [decimal proposal ](https://github.com/tc39/proposal-decimal) in the works as well! Or maybe they want to use a `bigint` for a specific conversation which couldn't be returned as a `number`. Maybe a `string` would work in that case but how would the library know that without being aware of the exact type at runtime. This is why I think it's important that the library returns the type of object that the wrapper is wrapping.

**Comment by bragle** [comment]
Sorry for the late reply - I completely forgot to follow up on this.

I've had a closer look at the wrapper approach, and while I probably won't be using it myself, I do see the value in making the library more flexible for people who need more precise numeric types or want to integrate with libraries, etc.

What I'm really interested in are the changes in the `alternative-number-support` branch. I think it's already a clear improvement. Since `v3-beta` already exists, would it make sense to include these changes there? That would make it easier to adopt and test, even before the docs, wrapper, etc. are finalized.

**Reply by Taar** [reply]
Sorry for the late reply, I believe the wrapper is in a good enough place to be released so I'll look into merging it into the main branch sometime this week.

I was already planning to completely replace the current docs with something that is a lot more maintainable so the docs for the wrapper (and examples) will be included within those changes. I'm hoping to start working on those changes this week.

**Reply by bragle** [reply]
Any updates on this? Happy to help if needed!

**Reply by Taar** [reply]
Sorry for the delay, I've included the feature in the `3.0.0-beta.8` release which has now been published to npm. Try it out and let me know if you experience any issues :) The docs don't contain any information about this feature (hopefully I'll find the time to add it in soon. I've been wanting to rework the docs for a while now) so feel free to post any questions about how to use it here. The tests contain an example using `decimal.js`.
